src/Widgets/VPN/VPNInfoDialog.vala

```diff
@@ -0,0 +1,161 @@
+        var image = new Gtk.Image.from_icon_name ("network-vpn", Gtk.IconSize.DIALOG);
+        image.hexpand = true;
+        image.halign = Gtk.Align.CENTER;
+
+        var data_grid = new Gtk.Grid ();
+        data_grid.expand = true;
+        data_grid.halign = Gtk.Align.START;
+
+        var state_label = new Gtk.Label (state);
+        state_label.selectable = true;
+        state_label.xalign = 0;
+
+        vpn_type = new Gtk.Label ("");
+        vpn_type.selectable = true;
+        vpn_type.xalign = 0;
+        vpn_type.no_show_all = true;
+
+        username = new Gtk.Label ("");
+        username.selectable = true;
+        username.xalign = 0;
+        username.no_show_all = true;
+
+        gateway = new Gtk.Label ("");
+        gateway.selectable = true;
+        gateway.xalign = 0;
+        gateway.no_show_all = true;
+
+        data_grid.attach (new VPNInfoLabel (_("Status: ")), 0, 0, 1, 1);
+        data_grid.attach (state_label, 1, 0, 2, 1);
+
+        data_grid.attach (new VPNInfoLabel (_("VPN Type: ")), 3, 0, 1, 1);
+        data_grid.attach (vpn_type, 4, 0, 2, 1);
+
+        data_grid.attach (new VPNInfoLabel (_("Username: ")), 0, 1, 1, 1);
+        data_grid.attach (username, 1, 1, 2, 1);
+
+        data_grid.attach (new VPNInfoLabel (_("Gateway: ")), 0, 2, 1, 1);
+        data_grid.attach (gateway, 1, 2, 5, 1);
+
+        data_grid.show_all ();
+
+
+        var grid = new Gtk.Grid ();
+        grid.expand = true;
+        grid.margin_start = grid.margin_end = 12;
+        grid.column_spacing = 6;
+        grid.orientation = Gtk.Orientation.VERTICAL;
+        grid.row_spacing = 6;
+
+        grid.add (image);
+        grid.add (data_grid);
+        grid.show_all ();
+
+        get_content_area ().add (grid);
+
+        add_button (_("_Close"), Gtk.ResponseType.CLOSE);
+    }
+
+    public void set_connection (NM.RemoteConnection _connection) {
+        connection = _connection;
+        connection.changed.connect (update_status);
+        update_status ();
+    }
+
+    //  From https://github.com/GNOME/gnome-control-center/blob/master/panels/network/net-vpn.c
+    private string get_key_group_username () {
+        switch (service_type) {
+            case "openvpn":
+            case "openconnect":
+                return "username";
+            case "vpnc":
+                return "Xauth username";
+            case "pptp":
+                return "user";
+            case "openswan":
+                return "leftxauthusername";
+        }
+
+        return "";
+    }
+
+    private string get_key_gateway () {
+        switch (service_type) {
+            case "openvpn":
+                return "remote";
+            case "vpnc":
+                return "IPSec gateway";
+            case "pptp":
+            case "openconnect":
+                return "gateway";
+            case "openswan":
+                return "right";
+        }
+
+        return "";
+    }
+
+
+    private string get_service_type () {
```

Comment:
I personally don't like that this method doesn't check if the `connection` isn't null, I think it would be better to make this static, take a connection argument and then call this from `update_status`.